api/observability/v1/input_types.go

```diff
@@ -0,0 +1,294 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// InputType specifies the type of log input to create.
+//
+// +kubebuilder:validation:Enum:=audit;application;infrastructure;receiver
+type InputType string
+
+// Reserved input names.
+const (
+	// InputTypeApplication contains all the non-infrastructure container logs.
+	InputTypeApplication InputType = "application"
+	// InputTypeInfrastructure contains infrastructure containers and system logs.
+	InputTypeInfrastructure InputType = "infrastructure"
+	// InputTypeAudit contains system audit logs.
+	InputTypeAudit InputType = "audit"
+	// InputTypeReceiver defines a network receiver for receiving logs from non-cluster sources.
+	InputTypeReceiver InputType = "receiver"
+)
+
+var (
+	InputTypes = []InputType{
+		InputTypeApplication,
+		InputTypeInfrastructure,
+		InputTypeAudit,
+		InputTypeReceiver,
+	}
+)
+
+// InputSpec defines a selector of log messages for a given log type.
+type InputSpec struct {
+	// Name used to refer to the input of a `pipeline`.
+	//
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+
+	// Type of output sink.
+	//
+	// +kubebuilder:validation:Required
+	Type InputType `json:"type"`
+
+	// Application, named set of `application` logs that
+	// can specify a set of match criteria
+	//
+	// +kubebuilder:validation:Optional
+	// +nullable
+	Application *Application `json:"application,omitempty"`
+
+	// Infrastructure, Enables `infrastructure` logs.
+	//
+	// +kubebuilder:validation:Optional
+	Infrastructure *Infrastructure `json:"infrastructure,omitempty"`
+
+	// Audit, enables `audit` logs.
+	//
+	// +kubebuilder:validation:Optional
+	Audit *Audit `json:"audit,omitempty"`
+
+	// Receiver to receive logs from non-cluster sources.
+	//
+	// +kubebuilder:validation:Optional
+	Receiver *ReceiverSpec `json:"receiver,omitempty"`
```

Comment:
You could use CEL to prevent the wrong union member from being added when the Type specifies a specific member, eg https://github.com/openshift/api/blob/ba11c1587003dc84cb014fd8db3fa597a3faaa63/example/v1/types_stable.go#L108-L125